forc-test/src/lib.rs

```diff
@@ -80,8 +102,163 @@ pub struct Opts {
+        let tests = pkg_with_tests
+            .entries
+            .iter()
+            .filter(|entry| entry.is_test())
+            .map(|entry| {
+                let offset =
+                    u32::try_from(entry.imm).expect("test instruction offset out of range");
+                let name = entry.fn_name.clone();
+                let test_setup = self.setup()?;
+                let (state, duration) = exec_test(&pkg_with_tests.bytecode, offset, test_setup);
+                let test_decl_id = entry
+                    .test_decl_id
+                    .clone()
+                    .expect("test entry point is missing declaration id");
+                let span = test_decl_id.span();
+                let test_function_decl = pkg_with_tests
+                    .declaration_engine
+                    .get_function(test_decl_id, &span)
+                    .expect("declaration engine is missing function declaration for test");
+                let condition = test_pass_condition(&test_function_decl)?;
+                Ok(TestResult {
+                    name,
+                    duration,
+                    span,
+                    state,
+                    condition,
+                })
+            })
+            .collect::<anyhow::Result<_>>()?;
+        let tested_pkg = TestedPackage {
+            built: Box::new(pkg_with_tests.clone()),
+            tests,
+        };
+        Ok(tested_pkg)
+    }
+
+    /// Setup the storage for a test and returns a contract id for testing contracts.
+    ///
+    /// For testing contracts, storage returned from this function contains the deployed contract.
+    /// For other types, default storage is returned.
+    fn setup(&self) -> anyhow::Result<TestSetup> {
+        match self {
+            PackageTests::Contract(contract_to_test) => {
+                let contract_pkg_without_tests = contract_to_test.tests_excluded.clone();
+                let test_setup = deploy_test_contract(contract_pkg_without_tests)?;
+                Ok(test_setup)
+            }
+            PackageTests::NonContract(_) => Ok(TestSetup {
+                storage: vm::storage::MemoryStorage::default(),
+                contract_id: None,
+            }),
+        }
+    }
+}
+
+impl Distribution<TxMetadata> for Standard {
+    /// Samples a random sample for `TxMetadata` which contains both random and constant variables.
+    /// For random variables a random sampling is done. For constant fields a constant value that
+    /// can be used directly with `TransactionBuilder` (for test transactions) is set.
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> TxMetadata {
+        TxMetadata {
+            secret_key: rng.gen(),
+            utxo_id: rng.gen(),
+            amount: 1,
+            asset_id: rng.gen(),
+            tx_pointer: rng.gen(),
+            maturity: 1,
+            block_height: (u32::MAX >> 1) as u64,
+        }
+    }
+}
```

Comment:
Rather than implementing `Distribution` here, we should probably use a dedicated constructor function. We don't randomise *all* fields of the `TxMetadata` here, I think implementing `Distribution` is a little dubious.